### ISSUE TITLE
IO reading process means you can't use execjs with V8

### ISSUE THREAD

**Issue by catmando** [opal/opal#1539]
execjs uses the process io stream to get the results from running an expression.

But opal detects the presence of the process object, and if it is there uses it rather than console.log for doing output via `puts` etc.

This means that if you are using execjs + V8 you can't do a `puts` !!!

I would favor reversing the order of the check so that if console is available, we use that, then if its not available we use the process object if its available, and finally if neither object is available, then we dump the output.

The problem is I can see this might be a breaking change.

Is there any way to have a patch that require a different approach while the former is deprecated?  I tried some simplistic ways to override the implementation, but I couldn't get it to stick.

**Comment by meh** [comment]
Another problem I can see happening is `console.log` *might* pretty print and do other fancy things, so you would end up finding yourself in the opposite situation, i.e. you want to print raw to stdout but you're pretty printing.

**Comment by catmando** [comment]
you are right... the issue is we are assuming so sort of "priority" which in some situation is the wrong choice.  So we just need a standard way to override.

**Comment by Mogztter** [comment]
We are also experiencing similar issues with Asciidoctor.js (see [opal_ext.rb](https://github.com/asciidoctor/asciidoctor.js/blob/master/lib/asciidoctor/opal_ext.rb#L9) and https://github.com/asciidoctor/asciidoctor.js/pull/190#issuecomment-228545983).

The main issue is that there's a variety of JavaScript environments/context:

 * Browser (V8)
 * Electron (Node)
 * Node.js
 * Nashorn (Java)
 * Rhino (Java)
 * SpiderMonkey
 * WebWorker
 * ExecJS
 * ...

I think we should be more specific and use fallback.

The following code should be enabled only if `process` is defined and `process.stdout` is defined and `process.stdout.write` methods exists, otherwise we should use the default method (ie. `console.log`):

````ruby
STDOUT.write_proc = -> (string) {`process.stdout.write(string)`}
````

So instead of detecting JavaScript environments, we should detect "features" and enable them accordingly. Otherwise every time a JavaScript environment comes out, we will need to tweak the code.

> So we just need a standard way to override.

I think this is a good idea! :+1:

**Comment by catmando** [comment]
In this case the problem is BOTH the V8 engine and React-Rails gem are providing the feature, but we want to use the one provided by React-rails NOT v8 engine.

Again I think its easy to solve by selectively requiring a file, but I can't quite figure out how to reset STDOUT.write_proc... if anybody could help me with that, I can propose a PR.

**Comment by elia** [comment]
The default write_proc can live in a constant or a global (kinda like `STDOUT` / `$stdout`), then should be fairly easy to do that. As I see it the default behavior should be a lightweight detection (that of course can be improved) and more specific stuff shuold be done in [`stdlib/opal/platform.rb`](https://github.com/opal/opal/blob/b8d72548d69238402eabc413a993a0a9baf1c51f/stdlib/opal/platform.rb).

Transitioning that to feature detection is fine by me as long as the semantics of requiring it in order to have the right stuff for your platform are preserved.

The current lightweight detection goes like this [as of now](https://github.com/opal/opal/blob/b8d72548d69238402eabc413a993a0a9baf1c51f/stdlib/opal/platform.rb):

```rb
STDOUT.write_proc = `typeof(process) === 'object' ? function(s){process.stdout.write(s)} : function(s){console.log(s)}`
STDERR.write_proc = `typeof(process) === 'object' ? function(s){process.stderr.write(s)} : function(s){console.warn(s)}`
```

Any proposal on how to improve is welcome 😃
